### ISSUE TITLE
Implement Pipeline with Modular Components Using Scikit-Learn

### ISSUE THREAD

**Issue by williamteles** [IoTDataAtelier/mlex#3]
## Description  
Develop a modular pipeline interface that integrates distinct components using Scikit-learn's `Pipeline` architecture. The goal is to clearly separate responsibilities and promote flexibility. The core modules to be implemented are:

- **DataProcessingStrategy**: Manages data ingestion, pre-processing, splitting, and sequencing.
- **ModelModule**: Wraps PyTorch-based models in a Scikit-learn-compatible interface.
- **EvaluationModule**: Provides standardized evaluation metrics and performance logging.
- **FeatureImportanceModule**: Performs feature importance analysis to interpret model decisions.

### 1. DataProcessingStrategy  
- **Purpose**: To encapsulate all data handling tasks, including reading raw data, cleaning, pre-processing, splitting, and sequencing if required.
- **Responsibilities**:
  - Apply transformations such as normalization, and encoding.
  - Split data into training and testing subsets.
  - Create sequences if the model requires sequential inputs.
- **Integration**: Serves as the first stage in the pipeline, ensuring the data is in the correct format for downstream modules.

### 2. ModelModule  
- **Purpose**: To integrate PyTorch models into the pipeline.
- **Responsibilities**:
  - Wrap PyTorch models (e.g., RNNs, LSTM, GRU) in a class that conforms to Scikit-learn's `BaseEstimator` interface.
  - Implement the standard methods (`fit`, `predict`, etc.) to facilitate training and inference.
  - Manage device configuration (CPU/GPU) and model state initialization.
- **Integration**: Acts as the core predictive component within the pipeline.

### 3. EvaluationModule  
- **Purpose**: To evaluate the performance of the models using standardized metrics.
- **Responsibilities**:
  - Calculate metrics such as accuracy, precision, recall, F1-score, and AUC-ROC.
  - Log evaluation results and provide summary reports.
  - Support comparative analysis between different model runs.
- **Integration**: Invoked after model predictions to assess and record performance.

### 4. FeatureImportanceModule  
- **Purpose**: To analyze and interpret model decisions through feature importance techniques.
- **Responsibilities**:
  - Implement methods such as SHAP, LIME, or permutation importance.
  - Generate visualizations and reports on feature contributions.
  - Provide insights for further feature engineering and model improvements.
- **Integration**: Complements the evaluation module by offering interpretability of the model's behavior.

## Implementation Steps

1. **Develop Modular Interfaces**  
   - Define abstract base classes or interfaces for each module to ensure consistency.
   - Implement concrete classes following the responsibilities outlined above.

2. **Integrate Modules into a Scikit-learn Pipeline**  
   - Use `sklearn.pipeline.Pipeline` to chain these modules together.
   - Ensure each module's output is compatible with the next module’s input.

3. **Examples**  
   - Include example scripts or Jupyter Notebooks demonstrating pipeline usage and configuration.

## Expected Outcome  
- A fully modular pipeline using Scikit-learn, with clearly defined components for data processing, model integration, evaluation, and feature importance.
- Enhanced maintainability and flexibility for future enhancements.
- Detailed documentation and examples to guide users in adapting and extending the pipeline.

**Comment by Gil32610** [comment]
Hey @williamteles @isadoravrx, check this out: https://skorch.readthedocs.io/en/stable/user/quickstart.html#in-an-sklearn-pipeline

```python
from skorch import NeuralNetClassifier
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

pipe.fit(X, y)
y_proba = pipe.predict_proba(X)

class MyModule(nn.Module):
    def __init__(self, num_units=10, nonlin=nn.ReLU()):
        super().__init__()
        #layer modules


    def forward(self, X, **kwargs):
        #any forward implementation


net = NeuralNetClassifier(
    MyModule,
    max_epochs=10,
    criterion=nn.CrossEntropyLoss(),
    lr=0.1,
    # Shuffle training data on each epoch
    iterator_train__shuffle=True,
)
pipe = Pipeline([
    ('scale', StandardScaler()),
    ('net', net),
])

pipe.fit(X, y)
y_proba = pipe.predict_proba(X)

```

**Comment by Gil32610** [comment]
Hey @williamteles @isadoravrx ! We need to talk about the Dataset and Dataloader implementation. I guess we would have to make some changes on the Base classes for the skorch pipeline to work properly.
